File: src/identity.js
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";
import { git } from "./process.js";
import { canonicalOrigin } from "./credentials.js";

export async function repositoryRoot(path = process.cwd()) {
  const result = await git(resolve(path), ["rev-parse", "--show-toplevel"], { allowFailure: true });
  if (result.code !== 0) throw new Error("this command must run inside a Git worktree");
  return result.stdout.trim();
}

async function localConfig(root, key) {
  const result = await git(root, ["config", "--local", "--get", key], { allowFailure: true });
  return result.code === 0 ? result.stdout.trim() : null;
}

export async function readIdentity(path = process.cwd()) {
  const root = await repositoryRoot(path);
  const service = await localConfig(root, "resync.service");
  const projectId = await localConfig(root, "resync.projectId");
  const checkoutId = await localConfig(root, "resync.checkoutId");
  if (!service && !projectId) return { root, service: null, projectId: null, checkoutId: null };
  if (!service || !projectId) throw new Error("incomplete RepoSync identity in local Git configuration");
  return { root, service: canonicalOrigin(service), projectId, checkoutId };
}

export async function writeIdentity(path, { service, projectId, checkoutId = `chk_${randomUUID().replaceAll("-", "")}` }) {
  const root = await repositoryRoot(path);
  const origin = canonicalOrigin(service);
  await git(root, ["config", "--local", "resync.service", origin]);
  await git(root, ["config", "--local", "resync.projectId", projectId]);
  await git(root, ["config", "--local", "resync.checkoutId", checkoutId]);
  return { root, service: origin, projectId, checkoutId };
}

export async function clearIdentity(path) {
  const root = await repositoryRoot(path);
  for (const key of ["resync.service", "resync.projectId", "resync.checkoutId"]) {
    await git(root, ["config", "--local", "--unset-all", key], { allowFailure: true });
  }
}

export function sameProject(left, right) {
  return Boolean(left?.service && right?.service && left?.projectId && right?.projectId &&
    canonicalOrigin(left.service) === canonicalOrigin(right.service) && left.projectId === right.projectId);
}

File: src/identity.rs
use crate::credentials::canonical_origin;
use crate::process::{RunOptions, git};
use anyhow::{Result, bail};
use std::path::{Path, PathBuf};
use uuid::Uuid;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Identity {
    pub root: PathBuf,
    pub service: Option<String>,
    pub project_id: Option<String>,
    pub checkout_id: Option<String>,
}

pub fn repository_root(path: &Path) -> Result<PathBuf> {
    let options = RunOptions {
        allow_failure: true,
        ..RunOptions::default()
    };
    let result = git(path, ["rev-parse", "--show-toplevel"], options)?;
    if result.code != 0 {
        bail!("this command must run inside a Git worktree");
    }
    Ok(PathBuf::from(result.stdout.trim()))
}

fn local_config(root: &Path, key: &str) -> Result<Option<String>> {
    let options = RunOptions {
        allow_failure: true,
        ..RunOptions::default()
    };
    let result = git(root, ["config", "--local", "--get", key], options)?;
    Ok((result.code == 0).then(|| result.stdout.trim().to_owned()))
}

pub fn read_identity(path: &Path) -> Result<Identity> {
    let root = repository_root(path)?;
    let service = local_config(&root, "resync.service")?;
    let project_id = local_config(&root, "resync.projectId")?;
    let checkout_id = local_config(&root, "resync.checkoutId")?;
    if service.is_some() != project_id.is_some() {
        bail!("incomplete RepoSync identity in local Git configuration");
    }
    Ok(Identity {
        root,
        service: service.map(|value| canonical_origin(&value)).transpose()?,
        project_id,
        checkout_id,
    })
}

pub fn write_identity(
    path: &Path,
    service: &str,
    project_id: &str,
    checkout_id: Option<&str>,
) -> Result<Identity> {
    let root = repository_root(path)?;
    let origin = canonical_origin(service)?;
    let checkout = checkout_id
        .map(str::to_owned)
        .unwrap_or_else(|| format!("chk_{}", Uuid::new_v4().simple()));
    for (key, value) in [
        ("resync.service", origin.as_str()),
        ("resync.projectId", project_id),
        ("resync.checkoutId", checkout.as_str()),
    ] {
        git(
            &root,
            ["config", "--local", key, value],
            RunOptions::default(),
        )?;
    }
    Ok(Identity {
        root,
        service: Some(origin),
        project_id: Some(project_id.into()),
        checkout_id: Some(checkout),
    })
}

pub fn clear_identity(path: &Path) -> Result<()> {
    let root = repository_root(path)?;
    for key in ["resync.service", "resync.projectId", "resync.checkoutId"] {
        let options = RunOptions {
            allow_failure: true,
            ..RunOptions::default()
        };
        git(&root, ["config", "--local", "--unset-all", key], options)?;
    }
    Ok(())
}

pub fn same_project(left: &Identity, right: &Identity) -> Result<bool> {
    Ok(
        match (
            &left.service,
            &left.project_id,
            &right.service,
            &right.project_id,
        ) {
            (Some(a), Some(b), Some(c), Some(d)) => {
                canonical_origin(a)? == canonical_origin(c)? && b == d
            }
            _ => false,
        },
    )
}
